cpp/src/arrow/compute/kernels/vector_dictionary.cc

```diff
@@ -0,0 +1,201 @@
+ public:
+  virtual Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                              ExecContext* ctx) const = 0;
+};
+
+template <typename IndiceArrowType>
+class DictionaryCompactKernelImpl : public DictionaryCompactKernel {
+  using BuilderType = NumericBuilder<IndiceArrowType>;
+  using CType = typename IndiceArrowType::c_type;
+
+ public:
+  Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                      ExecContext* ctx) const override {
+    const DictionaryArray& casted_dict_array =
+        checked_cast<const DictionaryArray&>(*dict_array);
+    const std::shared_ptr<Array>& dict = casted_dict_array.dictionary();
+    if (dict->length() == 0) {
+      return dict_array;
+    }
+    const std::shared_ptr<Array>& indices = casted_dict_array.indices();
+    if (indices->length() == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indices, empty_dict);
+    }
+    const CType* indices_data = indices->data()->GetValues<CType>(1);
+
+    // check whether the input is compacted
+    std::vector<bool> dict_used(dict->length(), false);
+    int64_t dict_used_count = 0;
+    for (int64_t i = 0; i < indices->length(); i++) {
+      if (indices->IsNull(i)) {
+        continue;
+      }
+
+      CType current_index = indices_data[i];
+      if (!dict_used[current_index]) {
+        dict_used[current_index] = true;
+        dict_used_count++;
+
+        if (dict_used_count == dict->length()) {  // input is already compacted
+          return dict_array;
+        }
+      }
+    }
+
+    // dictionary compaction
+    if (dict_used_count == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indices, empty_dict);
+    }
+    BuilderType dict_indices_builder(ctx->memory_pool());
+    bool need_change_indice = false;
+    CType len = static_cast<CType>(dict->length());
+    for (CType i = 0; i < len; i++) {
+      if (dict_used[i]) {
+        ARROW_RETURN_NOT_OK(dict_indices_builder.Append(i));
+      } else if (i + 1 < len && dict_used[i + 1]) {
+        need_change_indice = true;
+      }
+    }
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::Array> compacted_dict_indices,
+                          dict_indices_builder.Finish());
+    ARROW_ASSIGN_OR_RAISE(
+        auto compacted_dict_res,
+        Take(dict, compacted_dict_indices, TakeOptions::NoBoundsCheck(), ctx));
+    std::shared_ptr<arrow::Array> compacted_dict = compacted_dict_res.make_array();
+
+    // indices changes
+    if (!need_change_indice) {
+      return DictionaryArray::FromArrays(dict_array->type(), indices, compacted_dict);
+    }
+    std::vector<CType> indice_minus_number(dict->length(), 0);
```

Comment:
I think you can accomplish this more simply with another call to `Take`: just build an array mapping `compacted_dict_indices[i] -> i`

Reply:
Sorry, I don't get your idea. could you please describe it with more details?